sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs

```diff
@@ -0,0 +1,352 @@
+        public int MaxArraySize { get; }
+
+        /// <summary>
+        /// The backing array pool.
+        /// </summary>
+        public ArrayPool<byte> ArrayPool { get; }
+
+        /// <summary>
+        /// Absolute position of this stream in the larger stream it was chunked from.
+        /// </summary>
+        public long AbsolutePosition { get; }
+
+        /// <summary>
+        /// List of arrays making up the overall buffer. Since ArrayPool may give us a larger array than needed,
+        /// each array is paired with a count of the space actually used in the array. This <b>should</b> only
+        /// be important for the final buffer.
+        /// </summary>
+        private List<BufferPartition> BufferSet { get; } = new List<BufferPartition>();
+
+        private PooledMemoryStream(ArrayPool<byte> arrayPool, long absolutePosition, int maxArraySize)
+        {
+            AbsolutePosition = absolutePosition;
+            ArrayPool = arrayPool;
+            MaxArraySize = maxArraySize;
+        }
+
+        /// <summary>
+        /// Buffers a portion of the given stream, returning the buffered stream partition.
+        /// </summary>
+        /// <param name="stream">
+        /// Stream to buffer from.
+        /// </param>
+        /// <param name="minCount">
+        /// Minimum number of bytes to buffer. This method will not return until at least this many bytes have been read from <paramref name="stream"/> or the stream completes.
+        /// </param>
+        /// <param name="maxCount">
+        /// Maximum number of bytes to buffer.
+        /// </param>
+        /// <param name="absolutePosition">
+        /// Current position of the stream, since <see cref="Stream.Position"/> throws if not seekable.
+        /// </param>
+        /// <param name="arrayPool">
+        /// Pool to rent buffer space from.
+        /// </param>
+        /// <param name="maxArrayPoolRentalSize">
+        /// Max size we can request from the array pool.
+        /// </param>
+        /// <param name="async">
+        /// Whether to perform this operation asynchronously.
+        /// </param>
+        /// <param name="cancellationToken">
+        /// Cancellation token.
+        /// </param>
+        /// <returns>
+        /// The buffered stream partition with memory backed by an array pool.
+        /// </returns>
+        internal static async Task<PooledMemoryStream> BufferStreamPartitionInternal(
+            Stream stream,
+            long minCount,
+            long maxCount,
+            long absolutePosition,
+            ArrayPool<byte> arrayPool,
+            int? maxArrayPoolRentalSize,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            long totalRead = 0;
+            var streamPartition = new PooledMemoryStream(arrayPool, absolutePosition, maxArrayPoolRentalSize ?? DefaultMaxArrayPoolRentalSize);
+
+            int maxCountIndividualBuffer; // max count to write into buffer
```

Comment:
Super nit: can we put these comments on the line above?  Same for the rest of the comments in this file.